crates/factors-derive/src/lib.rs

```diff
@@ -0,0 +1,213 @@
+                Ok(())
+            }
+
+            pub fn configure_app(
+                &self,
+                app: #factors_path::App,
+                runtime_config: impl #factors_path::RuntimeConfigSource
+            ) -> #Result<#ConfiguredApp<Self>> {
+                let mut app_state = #app_state_name {
+                    #( #factor_names: None, )*
+                };
+                let mut runtime_config_tracker = #RuntimeConfigTracker::new(runtime_config);
+                #(
+                    app_state.#factor_names = Some(
+                        #Factor::configure_app(
+                            &self.#factor_names,
+                            #factors_path::ConfigureAppContext::<Self, #factor_types>::new(
+                                &app,
+                                &app_state,
+                                &mut runtime_config_tracker,
+                            )?,
+                        )?
+                    );
+                )*
+                runtime_config_tracker.validate_all_keys_used()?;
+                Ok(#ConfiguredApp::new(app, app_state))
+            }
+
+            pub fn build_store_data(&self, configured_app: &#ConfiguredApp<Self>, component_id: &str) -> #Result<#state_name> {
+                let app_component = configured_app.app().get_component(component_id).ok_or_else(|| {
+                    #wasmtime::Error::msg("unknown component")
+                })?;
+                let mut builders = #builders_name {
+                    #( #factor_names: None, )*
+                };
+                #(
+                    builders.#factor_names = Some(
+                        #Factor::prepare::<Self>(
+                            #factors_path::PrepareContext::new(
+                                &self.#factor_names,
+                                configured_app.app_state::<#factor_types>().unwrap(),
+                                &app_component,
+                            ),
+                            &mut #factors_path::InstanceBuilders::new(&mut builders),
+                        )?
+                    );
+                )*
+                Ok(#state_name {
+                    #(
+                        #factor_names: #FactorInstanceBuilder::build(
+                            builders.#factor_names.unwrap(),
+                        )?,
+                    )*
+                })
+            }
+
+        }
+
+        impl #factors_path::RuntimeFactors for #name {
+            type AppState = #app_state_name;
+            type InstanceBuilders = #builders_name;
+            type InstanceState = #state_name;
+
+            fn app_state<F: #Factor>(app_state: &Self::AppState) -> Option<&F::AppState> {
+                let type_id = #TypeId::of::<F>();
+                #(
+                    if type_id == #TypeId::of::<#factor_types>() {
+                        unsafe {
+                            return Some(::std::mem::transmute(&app_state.#factor_names));
```

Comment:
Hello! Long time user of `std::mem::transmute` here. There's a million subtle ways to get usage of `std::mem::transmute` wrong in ways that only sometimes show up or might break with future releases of the Rust compiler. From the docs:

> This makes transmute incredibly unsafe. transmute should be the absolute last resort.

I'm not convinced that we're in the situation where `transmute` is the only way to solve the issue at hand. Would it be possible to explore other solutions here?

Reply:
I tried using `Any::downcast_mut` in a previous iteration but couldn't get it working. That said, I _think_ this code is fundamentally doing the same thing as `Any::downcast_ref`: dynamically check the `TypeId` and then convert.

`Any::downcast_ref` converts via `as *const dyn Any as *const T` instead of `transmute`; not sure if there is any meaningful difference there.

Reply:
> I tried using Any::downcast_ref in a previous iteration but didn't get it working.

Apparently the trick was to "try harder": `<dyn Any>::downcast_ref` does the trick, replacing `transmute` here and in `instance_builder_mut` below.

There is still a `transmute` in `instance_state_offset`. That one is going to be trickier to avoid because it's used to implement borrow splitting; I don't have any other ideas for that one.